### PR TITLE
Fix multiple text rendering issues

### LINKED ISSUE #4806: TextBlock with Wrap cause an application freeze
When textblock with `TextWrapping="Wrap"` has a width less than one character width, this causes application to freeze.

2 different repros:

1.  very simple use anywhere in view `<TextBlock Text="Wrapped text" TextWrapping="Wrap" Width="3" />` application freezes when this view is shown
2. 
```xaml
    <Grid ColumnDefinitions="100,6,100">
      <TextBlock Text="Wrapped text" TextWrapping="Wrap" />
      <GridSplitter Grid.Column="1" />
      <TextBlock Grid.Column="2" Text="Wrapped text 2" TextWrapping="Wrap" />
    </Grid>
```
start sliding splitter towards the first TextBlock, application will freeze when it's too small to display one character.

repros has been tested with the current master branch

### PR DESCRIPTION
**Pull request by donandren**
## What does the pull request do?

This PR fixes multiple font rendering issues.

1. Make sure `TextWrapping.Wrap` is always producing a line with at least one character to avoid infinite loops
2. Make sure multiple TextRun's share the same baseline offset
3. Fix the sample TextBlockPage to properly show TextTrimming support with the new default font

## What is the current behavior?

1. It's possible to get infinite loops using `TextBlock` with `TextWraping.Wrap`, when the MaxWidth is less than the width of a character, thus causing a freeze.
2. Recent changes to `GlyphRun.BaselineOrigin` did not take into account that multiple text runs don't share the same font metrics and therefore are not aligned properly as a whole.

## What is the updated/expected behavior with this PR?

1. Fix text wrapping in situation with limited width (less than a character width)
2. Construct `ShapedTextCharacters` with a `GlyphRun` that has a `BaselineOrigin` of `(0,0)` so every run shares the same origin.

## How was the solution implemented (if it's not obvious)?

Text formatter doesn't return empty lines each line is at least one character
## Checklist

- [x] Added unit tests (if possible)?
- [x] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Avaloniaui.net with user documentation

## Fixed issues

Fixes #4806

### CODE REVIEW

**Comment by donandren** [comment]
@Gillibald pls can you take a look is that fix correct attempt for the #4806, thanks

File: src/Avalonia.Visuals/Media/TextFormatting/TextFormatterImpl.cs [hunk 1]
```diff
@@ -391,6 +391,10 @@ private static bool TryGetLineBreak(TextRun textRun, out LineBreak lineBreak)
                                 measuredLength = currentBreakPosition + lineBreaker.Current.PositionWrap;
                             }
                         }
+                        else if (measuredLength == 0 && currentLength == 0)
+                        {
+                            measuredLength = 1;
+                        }
```
Comment (donandren) [inline src/Avalonia.Visuals/Media/TextFormatting/TextFormatterImpl.cs:397]: @Gillibald pls can you take a look is that fix correct attempt for the #4806, thanks

**Review by Gillibald — COMMENTED** [review]
I think we should introduce a shortcut here: https://github.com/donandren/Avalonia/blob/14b26187987493b5ac048dac826e5453fcd79d9a/src/Avalonia.Visuals/Media/TextFormatting/TextFormatterImpl.cs#L353 to avoid redundant checks for line breaks in case we don't have WrapWithOverflow and a measured length of zero.

This probably needs some refactoring.

**Comment by Gillibald** [comment]
I have a branch in my fork that should have this fix included. 

https://github.com/Gillibald/Avalonia/tree/fixes/TextRendering

Feel free to apply my changes to TextFormatterImpl to your branch.

**Comment by donandren** [comment]
> 
> 
> I have a branch in my fork that should have this fix included.
> 
> https://github.com/Gillibald/Avalonia/tree/fixes/TextRendering
> 
> Feel free to apply my changes to TextFormatterImpl to your branch.

this fix also works great

**Review by Gillibald — APPROVED** [review]

**Pull request merged**